### PR TITLE
Create roles test in foundry and check admin roles granted on sft create

### PR DESCRIPTION
**Pull request by ninokeldishvili**

### CODE REVIEW

File: test/foundry/Roles.t.sol [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+import "../../contracts/test/TestErc20.sol";
+import "../../lib/openzeppelin-contracts-upgradeable/contracts/utils/StringsUpgradeable.sol";
+
+import "forge-std/Test.sol";
+import "forge-std/console.sol";
+
+import "../../contracts/vault/offchainAsset/OffchainAssetReceiptVaultFactory.sol";
+import "../../contracts/vault/offchainAsset/OffchainAssetReceiptVault.sol";
+
+contract OffChainAssetReceiptVaultTest is Test {
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:15]: you haven't deleted any hardhat tests in this PR @ninokeldishvili

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: test/foundry/Roles.t.sol [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        // Call the certify function
+        vault.certify(certifyUntil, blockNum, forceUntil, data);
+
+        vm.stopPrank();
+    }
+
+    function testConfiscateWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:198]: we can fuzz the address here

File: test/foundry/Roles.t.sol [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        bytes memory data = abi.encodePacked("Certification data");
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:200]: abi encoding copies bytes to a whole new region of memory but you can just cast strings directly like `bytes("foo")`

File: test/foundry/Roles.t.sol [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        );
+        vm.expectRevert(bytes(errorMessage));
+
+        //snapshot
+        vault.snapshot(data);
+    }
+
+    function testCertifyWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:170]: can fuzz this

File: test/foundry/Roles.t.sol [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+        // Get the current block number
+        uint256 blockNum = block.number;
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:173]: no need to create a value that is just using an evm constant

File: test/foundry/Roles.t.sol [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+        // Set up expected parameters
+        uint256 certifyUntil = block.timestamp + 1000;
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:176]: you can fuzz the "until"

File: test/foundry/Roles.t.sol [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        );
+        vm.expectRevert(bytes(errorMessage));
+
+        //set Tier
+        vault.setERC1155Tier(address(TierV2TestContract), minTier, context, data);
+    }
+
+    function testSnapshotWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:150]: you can fuzz alice

File: test/foundry/Roles.t.sol [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        );
+        vm.expectRevert(bytes(errorMessage));
+
+        //set Tier
+        vault.setERC20Tier(address(TierV2TestContract), minTier, context, data);
+    }
+
+    function testSetERC1155TierWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:125]: can fuzz alice

File: test/foundry/Roles.t.sol [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+        bytes memory data = "";
+        uint8 minTier = 10;
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:128]: can fuzz this tier

File: test/foundry/Roles.t.sol [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+        //set Tier
+        vault.setERC20Tier(address(TierV2TestContract), minTier, context, data);
+    }
+
+    function testSetERC1155TierWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        bytes memory data = "";
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:127]: can fuzz this data

File: test/foundry/Roles.t.sol [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+        //set Tier
+        vault.setERC1155Tier(address(TierV2TestContract), minTier, context, data);
+    }
+
+    function testSnapshotWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        bytes memory data = "";
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:152]: can fuzz this data

File: test/foundry/Roles.t.sol [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        vault.setERC20Tier(address(TierV2TestContract), minTier, context, data);
+    }
+
+    function testSetERC1155TierWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        bytes memory data = "";
+        uint8 minTier = 10;
+        uint256[] memory context = new uint256[](0);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:129]: can fuzz this context

File: test/foundry/Roles.t.sol [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+        bytes memory receiptInformation = "";
+
+        vm.expectRevert(abi.encodeWithSignature("MinShareRatio(uint256,uint256)", shareRatio, 0));
+        vault.deposit(aliceAssets, bob, shareRatio, receiptInformation);
+        vm.stopPrank();
+    }
+
+    function testSetERC20TierWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:100]: can fuzz alice

File: test/foundry/Roles.t.sol [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+
+        bytes memory data = "";
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:102]: fuzz data

File: test/foundry/Roles.t.sol [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+        uint8 minTier = 10;
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:103]: fuzz min tier

File: test/foundry/Roles.t.sol [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+        uint256[] memory context = new uint256[](0);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:104]: fuzz context

File: test/foundry/Roles.t.sol [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+        assert(HANDLER_ADMIN_Granted == true);
+        assert(ERC20TIERER_ADMIN_Granted == true);
+        assert(ERC1155TIERER_ADMIN_Granted == true);
+        assert(ERC20SNAPSHOTTER_ADMIN_Granted == true);
+        assert(CONFISCATOR_ADMIN_Granted == true);
+    }
+
+    function testDepositWithoutDepositorRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:86]: fuzz alice

File: test/foundry/Roles.t.sol [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+        // Get the second signer address
+        address bob = vm.addr(2);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:89]: fuzz bob

File: test/foundry/Roles.t.sol [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+        uint256 aliceAssets = 10;
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:90]: fuzz assets

File: test/foundry/Roles.t.sol [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+        bytes memory receiptInformation = "";
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:91]: fuzz receipt information

File: test/foundry/Roles.t.sol [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+        bool DEPOSITOR_ADMIN_Granted = vault.hasRole(DEPOSITOR_ADMIN, alice);
+        bool WITHDRAWER_ADMIN_Granted = vault.hasRole(WITHDRAWER_ADMIN, alice);
+        bool CERTIFIER_ADMIN_Granted = vault.hasRole(CERTIFIER_ADMIN, alice);
+        bool HANDLER_ADMIN_Granted = vault.hasRole(HANDLER_ADMIN, alice);
+        bool ERC20TIERER_ADMIN_Granted = vault.hasRole(ERC20TIERER_ADMIN, alice);
+        bool ERC1155TIERER_ADMIN_Granted = vault.hasRole(ERC1155TIERER_ADMIN, alice);
+        bool ERC20SNAPSHOTTER_ADMIN_Granted = vault.hasRole(ERC20SNAPSHOTTER_ADMIN, alice);
+        bool CONFISCATOR_ADMIN_Granted = vault.hasRole(CONFISCATOR_ADMIN, alice);
+
+        assert(DEPOSITOR_ADMIN_Granted == true);
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:74]: can `assertTrue` for all these

File: test/foundry/Roles.t.sol [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+    ReceiptFactory receiptFactory;
+    ReceiptVaultFactoryConfig factoryConfig;
+    OffchainAssetReceiptVault vault;
+    TestErc20 testErc20Contract;
+    ReadWriteTier TierV2TestContract;
+    address alice;
+    //shareRatio 1
+    uint256 shareRatio = 1e18;
+
+    function setUp() public {
```
Comment (thedavidmeister) [inline test/foundry/Roles.t.sol:29]: this all looks like it could be from an abstract contract

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: test/foundry/Roles.t.sol (hunk 3 above)
Comment (ninokeldishvili) [reply test/foundry/Roles.t.sol:185]: I'll consider that in future, but actually i fuzzed the data here

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

**Review by thedavidmeister — APPROVED** [review]

**Pull request merged**